src/components/Form/Widgets/FileWidget.tsx

```diff
@@ -1,248 +1,429 @@
+// we can add more as we get the usecases
+const getFileTypeIcon = (fileType: string) => {
+	switch (fileType) {
+		case 'image/png':
+		case 'image/jpg':
+		case 'image/jpeg':
+			return faFileImage;
+		default:
+			return faFile;
+	}
+};
+
+const fileSizeToReadable = (size: number) => {
+	const i = Math.floor(Math.log(size) / Math.log(1000));
+	return `${(size / Math.pow(1000, i)).toFixed(2)}${
+		['B', 'kB', 'MB', 'GB', 'TB'][i]
+	}`;
+};
+
+type AcceptedFile = File & {
+	loadingPercentage?: number;
+	base64Data?: string;
+	errors?: Array<{ code: ErrorCode | undefined; message: string }>;
+};
 
-export interface OnFileReadSuccessParams {
+// Improve error messages from react-dropzone
+const getErrorMessage = (
+	code: ErrorCode | undefined,
+	message: string,
+	maxSize: number,
+) => {
+	switch (code) {
+		case 'file-too-large':
+			return `File is larger than ${fileSizeToReadable(maxSize)}`;
+		default:
+			return message;
+	}
+};
+
+export interface OnFileReadParams {
 	dataUrl: string;
 	uploadedFile: File;
 }
 
-export interface OnFileReadErrorParams {
-	uploadErrors: string[];
-}
-
-// We need to forward uploadErrors onChange because RJSF does not support widget modifying
-// the default error handler.See: https://github.com/rjsf-team/react-jsonschema-form/issues/2718
-export type OnFileReadParams = OnFileReadSuccessParams | OnFileReadErrorParams;
-
 type FileWidgetProps = WidgetProps & {
+	// TODO: Currently it's not officially supported to raise errors via a custom widget's onChange, any attempt to do so would be a workaround
+	// See: https://github.com/rjsf-team/react-jsonschema-form/issues/2718
+	// There is a PR open to address this, so we should keep an eye on it and revisit this widget's "error-raising" when it's merged
+	// See: https://github.com/rjsf-team/react-jsonschema-form/pull/4188
 	onChange: (params?: OnFileReadParams) => void;
 	value?: OnFileReadParams | string;
 };
 
 export const FileWidget = ({
 	onChange,
-	value,
 	multiple,
-	...props
+	schema,
+	uiSchema,
 }: FileWidgetProps) => {
-	const { schema, uiSchema } = props;
+	const [files, setFiles] = useState<AcceptedFile[]>([]);
 
-	const [loadingPercentage, setLoadingPercentage] = useState<
-		number | undefined
-	>(undefined);
-	const [errors, setErrors] = useState<string[]>([]);
+	const fileUploadId = useMemo(() => `file-upload-${crypto.randomUUID()}`, []);
 
 	const accept = uiSchema?.['ui:options']?.accept as Accept | undefined;
 	const maxSize = uiSchema?.['ui:options']?.maxSize as number | undefined;
-
-	const dataUrl = (
-		dataUrlValue: OnFileReadParams | string,
-	): string | undefined => {
-		if (typeof dataUrlValue === 'string') {
-			return dataUrlValue;
-		}
-
-		if ('dataUrl' in dataUrlValue) {
-			return dataUrlValue.dataUrl;
-		}
-	};
-
-	const isImage = (value: OnFileReadParams | string) => {
-		return dataUrl(value)?.startsWith('data:image/');
-	};
+	const disabled = uiSchema?.['ui:options']?.disabled as boolean | undefined;
```

Comment:
To ensure clarity, should we add a `disabledReasons` property and a tooltip to explain to the user why drag-and-drop is disabled?

Also, the cast here is not necessary

```suggestion
	const disabled = uiSchema?.['ui:options']?.disabled;
```